policy/daemon.go

```diff
@@ -223,39 +238,39 @@ func (p *policyd) CheckPolicy(ctx context.Context, domain string, roles []string
 						glg.Debugf("Checking policy domain: %s, role: %v, action: %s, resource: %s, assertion: %v", domain, roles, action, resource, ass)
 						select {
 						case <-cctx.Done():
-							ch <- cctx.Err()
+							ch <- roleEffect{Role: role, Effect: cctx.Err()}
 							return
 						default:
 							// deny policies come first in rolePolicies, so it will return first before allow policies is checked
 							if strings.EqualFold(ass.ResourceDomain, domain) &&
 								ass.ActionRegexp.MatchString(strings.ToLower(action)) &&
 								ass.ResourceRegexp.MatchString(strings.ToLower(resource)) {
-								ch <- ass.Effect
+								ch <- roleEffect{Role: role, Effect: ass.Effect}
 								return
 							}
 						}
 					}
 				}
-			}(ech)
+			}(role, ech)
 		}
 		wg.Wait()
 	}()
 
-	allowed := false
-	for err := range ech {
-		if err != nil { // denied assertion is prioritize, so return directly
-			glg.Debugf("check policy domain: %s, role: %v, action: %s, resource: %s, result: %v", domain, roles, action, resource, err)
-			return err
+	var allowedRoles []string
+	for re := range ech {
+		if re.Effect != nil { // denied assertion is prioritize, so return directly
+			glg.Debugf("check policy domain: %s, role: %v, action: %s, resource: %s, result: %v", domain, roles, action, resource, re.Effect)
+			return nil, re.Effect
 		}
-		allowed = true
+		allowedRoles = append(allowedRoles, re.Role)
```

Comment:
Would be nice if the allowedRoles were allocated big enough to hold all results to prevent re-allocation

Reply:
Done